Src/Support/GoogleApis.Auth.PlatformServices_Shared/OAuth2/LocalServerCodeReceiver.cs

```diff
@@ -0,0 +1,393 @@
+                }
+                var queryParams = pathParts[1];
+                var result = queryParams.Split(new[] { '&' }, StringSplitOptions.RemoveEmptyEntries).Select(param =>
+                {
+                    var keyValue = param.Split('=');
+                    if (keyValue.Length > 2)
+                    {
+                        throw new ServerException($"Invalid query parameter: '{keyValue}'");
+                    }
+                    var key = WebUtility.UrlDecode(keyValue[0]);
+                    var value = keyValue.Length == 2 ? WebUtility.UrlDecode(keyValue[1]) : "";
+                    return new { key, value };
+                }).ToDictionary(x => x.key, x => x.value);
+                return result;
+            }
+
+            private async Task WaitForAllHeaders(Func<Task<byte?>> getByte)
+            {
+                // Looking for an empty line, terminated by \r\n
+                int byteCount = 0;
+                int lineLength = 0;
+                char c0 = (char)0;
+                char c1 = (char)0;
+                while (true)
+                {
+                    if (byteCount > MaxHeadersLength)
+                    {
+                        throw new ServerException($"Headers too long: > {MaxHeadersLength} bytes.");
+                    }
+                    byte? b = await getByte();
+                    if (b == null)
+                    {
+                        throw new ServerException("Unexpected end of network stream waiting for headers.");
+                    }
+                    c0 = c1;
+                    c1 = (char)b;
+                    lineLength += 1;
+                    byteCount += 1;
+                    if (c0 == '\r' && c1 == '\n')
+                    {
+                        // End of line
+                        if (lineLength == 2)
+                        {
+                            return;
+                        }
+                        lineLength = 0;
+                    }
+                }
+            }
+
+            private async Task WriteResponse(NetworkStream stream, CancellationToken cancellationToken)
+            {
+                string fullResponse = $"HTTP/1.1 200 OK\r\n\r\n{ClosePageResponse}";
+                var response = Encoding.ASCII.GetBytes(fullResponse);
+                await stream.WriteAsync(response, 0, response.Length, cancellationToken);
+                await stream.FlushAsync(cancellationToken);
+            }
+
+            public void Dispose()
+            {
+                _cts.Cancel();
+                _listener.Stop();
+            }
+        }
+
+        private string redirectUri;
+        public string RedirectUri
+        {
+            get
+            {
+                if (!string.IsNullOrEmpty(redirectUri))
+                {
+                    return redirectUri;
+                }
+
+                return redirectUri = string.Format(LoopbackCallback, GetRandomUnusedPort());
```

Comment:
Presumably this has the problem that we could find an unused port, but then by the time we want to use it, it's in use. Any thoughts on the best way of avoiding that?